File: app/src/main/java/com/huanchengfly/tieba/post/activities/BaseActivity.java
package com.huanchengfly.tieba.post.activities;

import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.CallSuper;
import androidx.annotation.ColorInt;
import androidx.annotation.Keep;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;

import com.gyf.immersionbar.ImmersionBar;
import com.huanchengfly.tieba.post.BaseApplication;
import com.huanchengfly.tieba.post.R;
import com.huanchengfly.tieba.post.ui.theme.interfaces.ExtraRefreshable;
import com.huanchengfly.tieba.post.ui.theme.utils.ThemeUtils;
import com.huanchengfly.tieba.post.utils.AppPreferencesUtils;
import com.huanchengfly.tieba.post.utils.HandleBackUtil;
import com.huanchengfly.tieba.post.utils.SharedPreferencesUtil;
import com.huanchengfly.tieba.post.utils.ThemeUtil;
import com.huanchengfly.tieba.post.widgets.VoicePlayerView;
import com.huanchengfly.tieba.post.widgets.theme.TintToolbar;

import butterknife.ButterKnife;
import cn.jzvd.Jzvd;
import me.imid.swipebacklayout.lib.app.SwipeBackActivity;

import static com.huanchengfly.tieba.post.utils.ColorUtils.getDarkerColor;
import static com.huanchengfly.tieba.post.utils.ThemeUtil.SP_CUSTOM_TOOLBAR_PRIMARY_COLOR;
import static com.huanchengfly.tieba.post.utils.ThemeUtil.THEME_CUSTOM;
import static com.huanchengfly.tieba.post.utils.ThemeUtil.THEME_TRANSLUCENT;

public abstract class BaseActivity extends SwipeBackActivity implements ExtraRefreshable {
    public static final int NO_LAYOUT = -1;
    private TintToolbar mTintToolbar;
    private String oldTheme;
    private boolean activityRunning = true;
    private int customStatusColor = -1;

    protected int getLayoutId() {
        return NO_LAYOUT;
    }

    protected AppPreferencesUtils getAppPreferences() {
        return new AppPreferencesUtils(this);
    }

    @Override
    protected void onPause() {
        super.onPause();
        activityRunning = false;
        Jzvd.releaseAllVideos();
    }

    protected boolean showDialog(Dialog dialog) {
        if (isActivityRunning()) {
            dialog.show();
            return true;
        }
        return false;
    }

    public boolean isActivityRunning() {
        return activityRunning;
    }

    @Override
    protected void onStop() {
        super.onStop();
        VoicePlayerView.Manager.release();
    }

    public boolean isNeedImmersionBar() {
        return true;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        fixBackground();
        getDeviceDensity();
        BaseApplication.getInstance().addActivity(this);
        ThemeUtil.setTheme(this);
        oldTheme = ThemeUtil.getTheme(this);
        if (isNeedImmersionBar()) {
            refreshStatusBarColor();
        }
        if (getLayoutId() != NO_LAYOUT) {
            setContentView(getLayoutId());
            ButterKnife.bind(this);
        }
    }

    private void fixBackground() {
        ViewGroup decor = (ViewGroup) getWindow().getDecorView();
        ViewGroup decorChild = (ViewGroup) decor.getChildAt(0);
        decorChild.setBackgroundColor(Color.BLACK);
    }

    public void refreshUIIfNeed() {
        if (TextUtils.equals(oldTheme, ThemeUtil.getTheme(this)) &&
                !THEME_CUSTOM.equals(ThemeUtil.getTheme(this)) &&
                !THEME_TRANSLUCENT.equals(ThemeUtil.getTheme(this))) {
            return;
        }
        if (recreateIfNeed()) {
            return;
        }
        ThemeUtils.refreshUI(this, this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        activityRunning = true;
        refreshUIIfNeed();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        BaseApplication.getInstance().removeActivity(this);
    }

    public void exitApplication() {
        BaseApplication.getInstance().removeAllActivity();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                if (!HandleBackUtil.handleBackPress(this)) {
                    finish();
                }
                return true;
            case R.id.menu_exit:
                finish();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (mTintToolbar != null) {
            mTintToolbar.tint();
        }
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        if (mTintToolbar != null) {
            mTintToolbar.tint();
        }
        return true;
    }

    @Override
    public void setSupportActionBar(@Nullable Toolbar toolbar) {
        super.setSupportActionBar(toolbar);
        if (toolbar instanceof TintToolbar) {
            mTintToolbar = (TintToolbar) toolbar;
        }
    }

    @Override
    public void onBackPressed() {
        if (!HandleBackUtil.handleBackPress(this)) {
            super.onBackPressed();
        }
    }

    public void setTitle(String newTitle) {
    }

    public void setSubTitle(String newTitle) {
    }

    protected void getDeviceDensity() {
        DisplayMetrics metrics = new DisplayMetrics();
        getWindowManager().getDefaultDisplay().getMetrics(metrics);
        int width = metrics.widthPixels;
        int height = metrics.heightPixels;
        BaseApplication.ScreenInfo.EXACT_SCREEN_HEIGHT = height;
        BaseApplication.ScreenInfo.EXACT_SCREEN_WIDTH = width;
        float density = metrics.density;
        BaseApplication.ScreenInfo.DENSITY = metrics.density;
        BaseApplication.ScreenInfo.SCREEN_HEIGHT = (int) (height / density);
        BaseApplication.ScreenInfo.SCREEN_WIDTH = (int) (width / density);
    }

    protected ValueAnimator colorAnim(ImageView view, int... value) {
        ValueAnimator animator = ObjectAnimator.ofArgb(new ImageViewAnimWrapper(view), "tint", value);
        animator.setDuration(150);
        animator.setInterpolator(new AccelerateDecelerateInterpolator());
        return animator;
    }

    protected ValueAnimator colorAnim(TextView view, int... value) {
        ValueAnimator animator = ObjectAnimator.ofArgb(new TextViewAnimWrapper(view), "textColor", value);
        animator.setDuration(150);
        animator.setInterpolator(new AccelerateDecelerateInterpolator());
        return animator;
    }

    public void setCustomStatusColor(int customStatusColor) {
        this.customStatusColor = customStatusColor;
        refreshStatusBarColor();
    }

    public void refreshStatusBarColor() {
        if (THEME_TRANSLUCENT.equals(ThemeUtil.getTheme(this))) {
            ImmersionBar.with(this)
                    .transparentBar()
                    .init();
        } else {
            ImmersionBar immersionBar = ImmersionBar.with(this)
                    .fitsSystemWindowsInt(true, ThemeUtils.getColorByAttr(this, R.attr.colorBg))
                    .navigationBarColorInt(ThemeUtils.getColorByAttr(this, R.attr.colorNavBar))
                    .navigationBarDarkIcon(ThemeUtil.isNavigationBarFontDark(this));
            if (customStatusColor != -1) {
                immersionBar.statusBarColorInt(customStatusColor)
                        .autoStatusBarDarkModeEnable(true);
            } else {
                immersionBar.statusBarColorInt(calcStatusBarColor(this, ThemeUtils.getColorByAttr(this, R.attr.colorToolbar)))
                        .statusBarDarkFont(ThemeUtil.isStatusBarFontDark(this));
            }
            immersionBar.init();
        }
    }

    public static int calcStatusBarColor(Context context, @ColorInt int originColor) {
        boolean darkerStatusBar = true;
        if (THEME_CUSTOM.equals(ThemeUtil.getTheme(context)) && !SharedPreferencesUtil.get(context, SharedPreferencesUtil.SP_SETTINGS)
                .getBoolean(SP_CUSTOM_TOOLBAR_PRIMARY_COLOR, true)) {
            darkerStatusBar = false;
        } else if (ThemeUtil.getTheme(context).equals(ThemeUtil.THEME_WHITE)) {
            darkerStatusBar = false;
        } else if (!SharedPreferencesUtil.get(context, SharedPreferencesUtil.SP_SETTINGS).getBoolean("status_bar_darker", true)) {
            darkerStatusBar = false;
        }
        return darkerStatusBar ? getDarkerColor(originColor) : originColor;
    }

    @CallSuper
    @Override
    public void refreshGlobal(Activity activity) {
        if (isNeedImmersionBar()) {
            refreshStatusBarColor();
        }
        oldTheme = ThemeUtil.getTheme(this);
    }

    private boolean recreateIfNeed() {
        if ((ThemeUtil.isNightMode(this) && !ThemeUtil.isNightMode(oldTheme)) ||
                (!ThemeUtil.isNightMode(this) && ThemeUtil.isNightMode(oldTheme))) {
            recreate();
            return true;
        }
        if ((oldTheme.equals(THEME_TRANSLUCENT) && !THEME_TRANSLUCENT.equals(ThemeUtil.getTheme(this))) ||
                (THEME_TRANSLUCENT.equals(ThemeUtil.getTheme(this)) && !oldTheme.equals(THEME_TRANSLUCENT))) {
            recreate();
            return true;
        }
        return false;
    }

    @Override
    public void refreshSpecificView(View view) {
    }

    @Keep
    protected static class TextViewAnimWrapper {
        private TextView mTarget;

        public TextViewAnimWrapper(TextView view) {
            mTarget = view;
        }

        @ColorInt
        public int getTextColor() {
            return mTarget.getCurrentTextColor();
        }

        public void setTextColor(@ColorInt int color) {
            mTarget.setTextColor(color);
        }
    }

    @Keep
    protected static class ImageViewAnimWrapper {
        private ImageView mTarget;

        public ImageViewAnimWrapper(ImageView view) {
            mTarget = view;
        }

        public int getTint() {
            return mTarget.getImageTintList() != null ? mTarget.getImageTintList().getDefaultColor() : 0x00000000;
        }

        public void setTint(@ColorInt int color) {
            mTarget.setImageTintList(ColorStateList.valueOf(color));
        }
    }
}

File: app/src/main/java/com/huanchengfly/tieba/post/activities/BaseActivity.kt
package com.huanchengfly.tieba.post.activities

import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.app.Activity
import android.app.Dialog
import android.content.res.ColorStateList
import android.content.res.Configuration
import android.content.res.Resources
import android.graphics.Color
import android.os.Bundle
import android.text.TextUtils
import android.util.DisplayMetrics
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateDecelerateInterpolator
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.CallSuper
import androidx.annotation.ColorInt
import androidx.annotation.Keep
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import butterknife.ButterKnife
import cn.jzvd.Jzvd
import com.gyf.immersionbar.ImmersionBar
import com.huanchengfly.tieba.post.App
import com.huanchengfly.tieba.post.App.Companion.INSTANCE
import com.huanchengfly.tieba.post.R
import com.huanchengfly.tieba.post.activities.MainActivity.Companion.SP_SHOULD_SHOW_SNACKBAR
import com.huanchengfly.tieba.post.dataStore
import com.huanchengfly.tieba.post.putBoolean
import com.huanchengfly.tieba.post.ui.common.theme.interfaces.ExtraRefreshable
import com.huanchengfly.tieba.post.ui.common.theme.utils.ThemeUtils
import com.huanchengfly.tieba.post.ui.widgets.VoicePlayerView
import com.huanchengfly.tieba.post.ui.widgets.theme.TintToolbar
import com.huanchengfly.tieba.post.utils.AppPreferencesUtils
import com.huanchengfly.tieba.post.utils.DialogUtil
import com.huanchengfly.tieba.post.utils.HandleBackUtil
import com.huanchengfly.tieba.post.utils.ThemeUtil
import com.huanchengfly.tieba.post.utils.calcStatusBarColorInt
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlin.coroutines.CoroutineContext

abstract class BaseActivity : AppCompatActivity(), ExtraRefreshable, CoroutineScope {
    val job = Job()
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + job

    private var mTintToolbar: TintToolbar? = null
    private var oldTheme: String = ""

    var isActivityRunning = true
        private set
    private var customStatusColor = -1
    private var statusBarTinted = false

    val appPreferences: AppPreferencesUtils by lazy { AppPreferencesUtils.getInstance(this) }

    override fun onPause() {
        super.onPause()
        isActivityRunning = false
        Jzvd.releaseAllVideos()
    }

    //禁止app字体大小跟随系统字体大小调节
    override fun getResources(): Resources {
        val fontScale = appPreferences.fontScale
        val resources = super.getResources()
        if (resources.configuration.fontScale != fontScale) {
            val configuration = resources.configuration
            configuration.fontScale = fontScale
            resources.updateConfiguration(configuration, resources.displayMetrics)
        }
        return resources
    }

    protected fun showDialog(dialog: Dialog): Boolean {
        if (isActivityRunning) {
            dialog.show()
            return true
        }
        return false
    }

    fun showDialog(builder: AlertDialog.Builder.() -> Unit): AlertDialog {
        val dialog = DialogUtil.build(this)
            .apply(builder)
            .create()
        if (isActivityRunning) {
            dialog.show()
        }
        return dialog
    }

    override fun onStop() {
        super.onStop()
        VoicePlayerView.Manager.release()
    }

    open val isNeedImmersionBar: Boolean = true
    open val isNeedFixBg: Boolean = true
    open val isNeedSetTheme: Boolean = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (isNeedFixBg) fixBackground()
        getDeviceDensity()
        INSTANCE.addActivity(this)
        if (isNeedSetTheme) ThemeUtil.setTheme(this)
        oldTheme = ThemeUtil.getTheme()
        if (isNeedImmersionBar) {
            refreshStatusBarColor()
        }
        if (getLayoutId() != -1) {
            setContentView(getLayoutId())
            ButterKnife.bind(this)
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        getDeviceDensity()
    }

    private fun fixBackground() {
        val decor = window.decorView as ViewGroup
        val decorChild = decor.getChildAt(0) as ViewGroup
        decorChild.setBackgroundColor(Color.BLACK)
    }

    fun refreshUIIfNeed() {
        if (TextUtils.equals(oldTheme, ThemeUtil.getTheme()) &&
            ThemeUtil.THEME_CUSTOM != ThemeUtil.getTheme() &&
            !ThemeUtil.isTranslucentTheme()
        ) {
            return
        }
        if (recreateIfNeed()) {
            return
        }
        ThemeUtils.refreshUI(this, this)
    }

    override fun onResume() {
        super.onResume()
        isActivityRunning = true
        if (appPreferences.followSystemNight) {
            if (App.isSystemNight && !ThemeUtil.isNightMode()) {
                dataStore.putBoolean(SP_SHOULD_SHOW_SNACKBAR, true)
                ThemeUtil.switchToNightMode(this, false)
            } else if (!App.isSystemNight && ThemeUtil.isNightMode()) {
                dataStore.putBoolean(SP_SHOULD_SHOW_SNACKBAR, true)
                ThemeUtil.switchFromNightMode(this, false)
            }
        }
        refreshUIIfNeed()
    }

    override fun onDestroy() {
        super.onDestroy()
        INSTANCE.removeActivity(this)
        job.cancel()
    }

    fun exitApplication() {
        INSTANCE.removeAllActivity()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                if (!HandleBackUtil.handleBackPress(this)) {
                    finish()
                }
                return true
            }
            R.id.menu_exit -> {
                finish()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (mTintToolbar != null) {
            mTintToolbar!!.tint()
        }
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        if (mTintToolbar != null) {
            mTintToolbar!!.tint()
        }
        return true
    }

    override fun setSupportActionBar(toolbar: Toolbar?) {
        super.setSupportActionBar(toolbar)
        if (toolbar is TintToolbar) {
            mTintToolbar = toolbar
        }
    }

    override fun onBackPressed() {
        if (!HandleBackUtil.handleBackPress(this)) {
            super.onBackPressed()
        }
    }

    open fun setTitle(newTitle: String?) {}
    open fun setSubTitle(newTitle: String?) {}

    private fun getDeviceDensity() {
        val metrics = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(metrics)
        val width = metrics.widthPixels
        val height = metrics.heightPixels
        App.ScreenInfo.EXACT_SCREEN_HEIGHT = height
        App.ScreenInfo.EXACT_SCREEN_WIDTH = width
        val density = metrics.density
        App.ScreenInfo.DENSITY = metrics.density
        App.ScreenInfo.SCREEN_HEIGHT = (height / density).toInt()
        App.ScreenInfo.SCREEN_WIDTH = (width / density).toInt()
    }

    protected fun colorAnim(view: ImageView, vararg value: Int): ValueAnimator {
        val animator: ValueAnimator =
            ObjectAnimator.ofArgb(ImageViewAnimWrapper(view), "tint", *value)
        animator.duration = 150
        animator.interpolator = AccelerateDecelerateInterpolator()
        return animator
    }

    protected fun colorAnim(view: TextView, vararg value: Int): ValueAnimator {
        val animator: ValueAnimator =
            ObjectAnimator.ofArgb(TextViewAnimWrapper(view), "textColor", *value)
        animator.duration = 150
        animator.interpolator = AccelerateDecelerateInterpolator()
        return animator
    }

    fun setCustomStatusColor(customStatusColor: Int) {
        if (ThemeUtil.isTranslucentTheme()) {
            return
        }
        this.customStatusColor = customStatusColor
        refreshStatusBarColor()
    }

    open fun refreshStatusBarColor() {
        if (ThemeUtil.isTranslucentTheme()) {
            ImmersionBar.with(this)
                .transparentBar()
                .init()
        } else {
            ImmersionBar.with(this).apply {
                if (customStatusColor != -1) {
                    statusBarColorInt(customStatusColor)
                    autoStatusBarDarkModeEnable(true)
                } else {
                    statusBarColorInt(
                        calcStatusBarColorInt(
                            this@BaseActivity,
                            ThemeUtils.getColorByAttr(this@BaseActivity, R.attr.colorToolbar)
                        )
                    )
                    statusBarDarkFont(ThemeUtil.isStatusBarFontDark())
                }
                fitsSystemWindowsInt(
                    true,
                    ThemeUtils.getColorByAttr(this@BaseActivity, R.attr.colorBg)
                )
                navigationBarColorInt(
                    ThemeUtils.getColorByAttr(
                        this@BaseActivity,
                        R.attr.colorNavBar
                    )
                )
                navigationBarDarkIcon(ThemeUtil.isNavigationBarFontDark())
            }.init()
        }
        if (!statusBarTinted) {
            statusBarTinted = true
        }
    }

    @CallSuper
    override fun refreshGlobal(activity: Activity) {
        if (isNeedImmersionBar) {
            refreshStatusBarColor()
        }
        oldTheme = ThemeUtil.getTheme()
    }

    private fun recreateIfNeed(): Boolean {
        if (ThemeUtil.isNightMode() && !ThemeUtil.isNightMode(oldTheme) ||
            !ThemeUtil.isNightMode() && ThemeUtil.isNightMode(oldTheme)
        ) {
            recreate()
            return true
        }
        if (oldTheme.contains(ThemeUtil.THEME_TRANSLUCENT) &&
            !ThemeUtil.isTranslucentTheme() || ThemeUtil.isTranslucentTheme() &&
            !oldTheme.contains(ThemeUtil.THEME_TRANSLUCENT)
        ) {
            recreate()
            return true
        }
        return false
    }

    override fun refreshSpecificView(view: View) {}

    @Keep
    protected class TextViewAnimWrapper(private val mTarget: TextView) {
        @get:ColorInt
        var textColor: Int
            get() = mTarget.currentTextColor
            set(color) {
                mTarget.setTextColor(color)
            }
    }

    @Keep
    protected class ImageViewAnimWrapper(private val mTarget: ImageView) {
        var tint: Int
            get() = if (mTarget.imageTintList != null) mTarget.imageTintList!!.defaultColor else 0x00000000
            set(color) {
                mTarget.imageTintList = ColorStateList.valueOf(color)
            }
    }

    open fun getLayoutId(): Int = -1

    fun launchIO(
        start: CoroutineStart = CoroutineStart.DEFAULT,
        block: suspend CoroutineScope.() -> Unit
    ): Job {
        return launch(Dispatchers.IO + job, start, block)
    }
}